File: jotupix_demo_ios/JotupixDemo/Jotupix/Program/JTextDiyColorContent.m
#import "JTextDiyColorContent.h"
#import "JByteWriter.h"
#import "JColor.h"

@implementation JTextDiyColorContent

#pragma mark - Static Multicolor Data

+ (NSArray<NSNumber *> *)multicolorData {
    return @[
        @(JColorRed),
        @(JColorYellow),
        @(JColorGreen),
        @(JColorCyan),
        @(JColorBlue),
        @(JColorPurple)
    ];
}

+ (uint32_t)multicolorDataSize {
    return (uint32_t)[self multicolorData].count;
}

#pragma mark - Init

- (instancetype)init {
    if (self = [super init]) {
        self.contentType = JContentTypeDIYColor;
        _textData = [NSMutableArray array];
    }
    return self;
}

#pragma mark - Packing

- (NSData *)get {

    JByteWriter *content = [[JByteWriter alloc] init];

    // Content type
    [content put_u8:(uint8_t)self.contentType];

    // 5 bytes reserved
    [content put_repeat:0 count:5];

    [content put_u16:self.moveSpace];
    [content put_u16:self.showX];
    [content put_u16:self.showY];
    [content put_u16:self.showWidth];
    [content put_u16:self.showHeight];
    [content put_u8:self.showMode];
    [content put_u8:self.showSpeed];
    [content put_u8:self.stayTime];

    // Reserved 1 byte
    [content put_u8:0];

    [content put_u16:self.textNum];

    // content.put_u16(m_textAllWide & 0xFFFFF)
    [content put_u16:(self.textAllWide & 0xFFFFF)];

    // First loop: text width list
    for (JTextFont *font in self.textData) {
        [content put_u8:font.textWidth];
    }

    // Second loop: text color list
    for (JTextFont *font in self.textData) {
        [content put_u16:font.textColor];
    }

    // Insert total block size at beginning (include 4 bytes for size itself)
    uint32_t totalSize = (uint32_t)(content.size + 4);

    // Correct usage — insert at pos=0, value=totalSize
    [content insert_u32:0 value:totalSize];

    // Return NSData, not uint8_t*
    return [NSData dataWithBytes:content.buffer length:content.size];
}

@end

File: jotupix_demo_android/jotupix/src/main/java/com/jtkj/jotupix/program/JTextDiyColorContent.java
package com.jtkj.jotupix.program;

import com.jtkj.jotupix.utils.JByteWriter;

import java.util.ArrayList;
import java.util.List;

public class JTextDiyColorContent extends JContentBase {

    public static final int[] multicolorData = {
            JColor.RED,
            JColor.GREEN,
            JColor.BLUE,
            JColor.YELLOW,
            JColor.CYAN,
            JColor.PURPLE,
            JColor.WHITE
    };

    public int moveSpace;
    public int showX;
    public int showY;
    public int showWidth;
    public int showHeight;
    public int showMode;
    public int showSpeed;
    public int stayTime;
    public int textNum;
    public int textAllWide;

    public List<JTextFont> textData;

    public JTextDiyColorContent() {
        this.contentType = ContentType.DIYCOLOR;
        this.textData = new ArrayList<>();
    }

    @Override
    public byte[] get() {
        JByteWriter content=new JByteWriter();

        // Put content type
        content.putByteOne(contentType);

        // Put 5 bytes of zeros
        content.putRepeatByteOne(0, 5);

        // Put movespace
        content.putBytesTwo(moveSpace);

        // Put coordinates and dimensions
        content.putBytesTwo(showX);
        content.putBytesTwo(showY);
        content.putBytesTwo(showWidth);
        content.putBytesTwo(showHeight);

        // Put display properties
        content.putByteOne(showMode);
        content.putByteOne(showSpeed);
        content.putByteOne(stayTime);

        // Put default  1 byte zero
        content.putByteOne(0);

        // Put text properties
        content.putBytesTwo(textNum);
        content.putBytesTwo(textAllWide);

        // Put text widths
        for (JTextFont font : textData) {
            content.putByteOne(font.textWidth);
        }

        // Put text colors
        for (JTextFont font : textData) {
            content.putBytesTwo(font.textColor);
        }

        // Insert total size at the beginning (size + 4)
        content.insertBytesFour(0, content.size() + 4);

        return content.toByteArray();
    }

}